src/utils/weights_adapter.cc

```diff
@@ -25,19 +25,56 @@
   Program grant you additional permission to convey the resulting work.
 */
 
-#include "src/utils/weights_adapter.h"
+#include "utils/weights_adapter.h"
+
+#include <absl/base/optimization.h>
+
+#include "utils/bf16_utils.h"
+#include "utils/exception.h"
+#include "utils/fp16_utils.h"
 
 namespace lczero {
+
 float LayerAdapter::Iterator::ExtractValue(const uint16_t* ptr,
                                            const LayerAdapter* adapter) {
-  return *ptr / static_cast<float>(0xffff) * adapter->range_ + adapter->min_;
+  switch (adapter->encoding_) {
+    case pblczero::Weights::Layer::LINEAR16: {
+      float theta = *ptr / static_cast<float>(0xffff);
+      return adapter->min_ * (1 - theta) + adapter->max_ * theta;
+    }
+    case pblczero::Weights::Layer::FLOAT16:
+      return FP16toFP32(*ptr);
+    case pblczero::Weights::Layer::BFLOAT16:
+      return BF16toFP32(*ptr);
+    [[unlikely]] default:  // To silence a couple of warnings.
+#if defined(ABSL_UNREACHABLE)
+      ABSL_UNREACHABLE();
+#elif defined(__GNUC__)
```

Comment:
Do we need all that given that we #include <absl/base/optimization.h>?

Reply:
Unfortunately my Ubuntu 24.04 has an abseil version from 2022 while `ABSL_UNREACHABLE` was introduced in 2023.